tools/wasi/wasi.cpp

```diff
@@ -0,0 +1,235 @@
+uvwasi_t state;
+
+fizzy::execution_result wasi_return_enosys(fizzy::Instance&, std::vector<uint64_t>, int)
+{
+    return {false, {UVWASI_ENOSYS}};
+}
+
+fizzy::execution_result wasi_proc_exit(fizzy::Instance&, std::vector<uint64_t> args, int)
+{
+    uvwasi_proc_exit(&state, static_cast<uvwasi_exitcode_t>(args[0]));
+    // Should not reach this.
+    return {false, {}};
+}
+
+fizzy::execution_result wasi_fd_write(fizzy::Instance& instance, std::vector<uint64_t> args, int)
+{
+    uint32_t fd = static_cast<uint32_t>(args[0]);
+    uint32_t iovptr = static_cast<uint32_t>(args[1]);
+    uint32_t iovcnt = static_cast<uint32_t>(args[2]);
+    uint32_t nwritten = static_cast<uint32_t>(args[3]);
+
+    std::vector<uvwasi_ciovec_t> iovs(iovcnt);
+    // TODO: not sure what to pass as end, passing memory size...
+    uvwasi_errno_t ret = uvwasi_serdes_readv_ciovec_t(instance.memory->data(),
+        instance.memory->size(), iovptr, iovs.data(), static_cast<uvwasi_size_t>(iovs.size()));
+    if (ret != UVWASI_ESUCCESS)
+        return {false, {ret}};
+
+    uvwasi_size_t nwritten_;
+    ret = uvwasi_fd_write(&state, static_cast<uvwasi_fd_t>(fd), iovs.data(),
+        static_cast<uvwasi_size_t>(iovs.size()), &nwritten_);
+    uvwasi_serdes_write_uint32_t(instance.memory->data(), nwritten, nwritten_);
+
+    return {false, {ret}};
+}
+
+fizzy::execution_result wasi_fd_read(fizzy::Instance& instance, std::vector<uint64_t> args, int)
+{
+    uint32_t fd = static_cast<uint32_t>(args[0]);
+    uint32_t iovptr = static_cast<uint32_t>(args[1]);
+    uint32_t iovcnt = static_cast<uint32_t>(args[2]);
+    uint32_t nread = static_cast<uint32_t>(args[3]);
+
+    std::vector<uvwasi_iovec_t> iovs(iovcnt);
+    // TODO: not sure what to pass as end, passing memory size...
+    uvwasi_errno_t ret = uvwasi_serdes_readv_iovec_t(instance.memory->data(),
+        instance.memory->size(), iovptr, iovs.data(), static_cast<uvwasi_size_t>(iovs.size()));
+    if (ret != UVWASI_ESUCCESS)
+        return {false, {ret}};
+
+    uvwasi_size_t nread_;
+    ret = uvwasi_fd_read(&state, static_cast<uvwasi_fd_t>(fd), iovs.data(),
+        static_cast<uvwasi_size_t>(iovs.size()), &nread_);
+    uvwasi_serdes_write_uint32_t(instance.memory->data(), nread, nread_);
+
+    return {false, {ret}};
+}
+
+fizzy::execution_result wasi_fd_prestat_get(
+    fizzy::Instance& instance, std::vector<uint64_t> args, int)
+{
+    uint32_t fd = static_cast<uint32_t>(args[0]);
+    uint32_t prestat_ptr = static_cast<uint32_t>(args[1]);
+
+    uvwasi_prestat_t buf;
+    uvwasi_errno_t ret = uvwasi_fd_prestat_get(&state, fd, &buf);
+
+    uvwasi_serdes_write_prestat_t(instance.memory->data(), prestat_ptr, &buf);
+
+    return {false, {ret}};
+}
+
+fizzy::execution_result wasi_environ_sizes_get(
+    fizzy::Instance& instance, std::vector<uint64_t> args, int)
+{
+    uint32_t environc = static_cast<uint32_t>(args[0]);
+    // TODO: implement properly (only returns 0 now)
```

Comment:
This function should be returning two size values, if I'm looking at correct spec
https://github.com/WebAssembly/WASI/blob/4fa4f4a82844f1dd5e45cbcb7c714e602f6962de/phases/snapshot/witx/wasi_snapshot_preview1.witx#L39

Reply:
You are correct, fixed.